### DISCUSSION TITLE
G7 without Dexcom app / Authentication

### DISCUSSION THREAD

**Discussion by SBriere** [gui-dos/DiaBLE#17] [General]
Very interesting project! I've made a development setup (so not using TestFlight) in order to test the data retrieval directly from a Dexcom G7 sensor.

I notice that your latest commits seem related to dealing with the authentication process needed to connect directly using BLE to that sensor. Haven't been able to get any data from the device so far - it always disconnect just after the J-PAKE payload being received, but I suppose it's a normal thing in the current development state?

Or maybe am I missing something in the settings (such as "Sensor Code" and "Transmitter Serial" - which is always set to "XXXXkW")?

Also noticing that the "test" mode doesn't include authentication - I suppose it's meant to be used with the Dexcom app running in the background?

**Comment by gui-dos** [comment]
Yep: the third “Test” mode in the Console is intended for beta testers and those interested in eavesdropping the incoming BLE traffic to the official apps running in parallel, avoiding to send authentication commands that would cause sensors to disconnect.

Even if xDrip4iO5 developers don’t recommend it and make it available only among the Developer Settings, this  is a way for example to be able to decrypt the Libre 2 data running in parallel with other apps since only the sensor UID scannable via NFC is required, not the `patchInfo` at the time of the sensor activation! ;-)

And this is the current way that the Dexcom G7 support works in all the other iOS apps, requiring the official app to run in the background: differently than Abbott, luckily Dexcom is not encrypting the 5-minute realtime values and the backfill stream!

Dexcom wraps the mbedtls framework to implement the J-PAKE authentication protocol (https://github.com/Mbed-TLS/mbedtls/blob/development/tf-psa-crypto/drivers/builtin/include/mbedtls/ecjpake.h) also for the standalone Watch app and, since Android’s xDrip is implementing J-PAKE successfully, perhaps implementing it in iOS might be a reachable goal.

From the decrypted IPA of Tandem’s t-connect app (i.e. https://decrypt.day/app/id1455916023) you can notice that they are implementing J-PAKE in their own way, apparently without using a partner SDK black box but, if you search GitHub for a pure Swift implementation of J-PAKE, you don’t get any result. The clean implementation https://github.com/particle-iot/iOSBLEExample/blob/main/iOSBLEExample/ParticleBLECode/ECJPake.swift for example also leverages on their mbedtls wrapping.

The Dexcom G7 support implemented in https://github.com/LoopKit/G7SensorKit seems really limited and poorly documented, too tied to the past.

In the Dexcom G7, for example, there aren’t separate “Tx” and “Rx” opcodes: the same opcode is used as the header byte marker for both the request commands and the sensor replies. Most commands now don’t even require a final CRC, also, and countless detailed primary and secondary state codes are sent.

Specific commands for requesting backfilling data and further encrypted streams regarding internal calibration and factory data (and even a debugging backdoor) are clearly available.

I am a Libre 3 final user but, having to purchase Dexcoms on eBay on my own, I am equally motivated and intrigued to explore the Dexcom internals anyway: hopefully some more third party apps by the growing number of pump manufacturers partners will leak out  crucial details on the cryptographic algorithms…. ;-)

**Reply by ianjorgensen** [reply]
@gui-dos id be more than happy to ship you G7s to wherever you need them, I would love to be able to connect directly without the official app installed. 

I had to resintall the dexcom g7 app when it gave me an error message when trying to pair a new sensor while traveling abroad and dexcom didn't let me login or signup to pair a new device and get  readings. I was told to go by dexcom support back home to use the cgm again. This is know issue 🤯

**Reply by gui-dos** [reply]
@ianjorgensen

Pairing a Dexcom G7 sensor requires going through a multi-stage challenge-response protocol involving certificate exchanges and validations.

Similar to the Libre 3, the authentication commands follow a sequential order (`02 03 04 05 06 07 …`) therefore you can easily grasp the logic workflow of the activation process: https://github.com/gui-dos/DiaBLE/blob/a5b8c521c1c8ecfdb060db646ee1a40bf7d77e13/DiaBLE/DexcomG7.swift#L78-L119

I am already happy to be able to detect the different BLE characteristics involved and the amount and kind of data sent/received which  other DIY open-source projects seem to have no clue about and/or not willing to document… B-)

**Reply by ps2** [reply]
@gui-dos sent you a note on Loop zulip, if you have a moment.

**Reply by gui-dos** [reply]
@ps2

I am using PacketLogger included in the Xcode Additional Tools to sniff the Libre/Dexcom BLE traffic.

I posted, for example, in the comments to https://github.com/gui-dos/DiaBLE/commit/7d4deeb771f53d04bb97bd0f690b23edb996c5d9, beside the zipped logs captured during the initial J-PAKE pairing and the later backfilling traffic, even a screenshot of my standalone watchOS implementation eavesdropping the official app running in parallel, but I got no feedback. Frustrated, I unsubscribed and rejoined the Zulip chat, but I was happy with my lonely world-exclusive scoop anyway ;-)

I am not rooting/jailbreaking my daily Pixels/iPhones, but I have a licensed copy of Hopper, and I am fond of comparing the decompiled Android bytecode by jadx-gui with the Swift struct descriptors where the enum constants’ definitions are not sorted alphabetically but in the correct camelCase order! B-)

I met a very responsive Chinese bot who provides decrypted IPAs (with https://github.com/ChiChou/bagbak) for using with PlayCover on Silicon Macs. I am probably the guy who is populating  the medical category https://decrypt.day/apps/medical the most: B-) (the Stelo app has been uncategorized on https://decrypt.day/app/id6475610406).

In the past, I succeeded in importing the US IPA and using the ONE/G7s I bought on eBay before they were available in Italy, but, you know, regulations and distribution channels can vary a lot from country to country, and I’m not sure if the Stelo would be available here anytime soon... :-(

I just bought a couple of ONE+s which are sold here quite cheaply for 42€ not by Dexcom but by Roche/AccuCheck (https://shop.dexcomone.info/eShop/monitoraggioincontinuo): upon looking at the IPA, the support for Direct-To-Watch looks like it is already available behind geofencing... ;-)

**Reply by cachrisman** [reply]
Hi @gui-dos — I'm a contributor to [Trio](https://github.com/nightscout/Trio) and I've been investigating how to get fresh CGM data to the Trio watch complication independently of WatchConnectivity. Your work here has been incredibly valuable — this thread is one of the most useful primary sources I've found.

I wanted to share what I found and ask a few questions.

**Inspecting the Dexcom G7 IPA**

I got the Dexcom G7 IPA from decrypt.day and extracted the watch extension entitlements via codesign. The `G7Watch Extension.appex` entitlements are:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "https://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>application-identifier</key>
    <string>6WC99N3397.com.dexcom.g7app.watchkitapp.watchkitextension</string>
    <key>com.apple.developer.bluetooth-central-background</key>
    <true />
    <key>com.apple.developer.bluetooth-central-screen-off-scanning</key>
    <true />
    <key>com.apple.developer.team-identifier</key>
    <string>R8894QKYB2</string>
    <key>com.apple.developer.usernotifications.critical-alerts</key>
    <true />
    <key>com.apple.security.application-groups</key>
    <array>
      <string>group.com.dexcom.g7app.watchkitapp</string>
    </array>
  </dict>
</plist>
```

That's it. No `HKWorkoutSession`, no `WKExtendedRuntimeSession`, no health-monitoring entitlement. The sustained background BLE is entirely explained by those two private entitlements. An Apple Engineer confirmed on the developer forums in 2019 that `bluetooth-central-background` is specifically intended for watchOS apps communicating with CGM devices: https://developer.apple.com/forums/thread/109947

**Open questions I still have**

1. Has anyone in the open-source CGM/APS community ever tried requesting `com.apple.developer.bluetooth-central-background` from Apple? I've submitted a request through the Apple developer portal but that was just today and I haven't heard back yet. I'd love to know if anyone has tried this path before.

2. Your README mentions that keeping the G7 connection alive in the background on watchOS "will require additional work" — was that written before the watchOS 9 WWDC22 background BLE improvements, or have you tested those and found them insufficient? The 5-slot timely alerts budget seems like it wouldn't be enough for continuous CGM delivery, but I'm wondering if your assessment was based on that or something else.

3. You mentioned a screenshot of your standalone watchOS implementation eavesdropping the Dexcom watch app running in parallel. Can you confirm that test was run with the Dexcom watch app actively connected at the same time? That would confirm CoreBluetooth reference-counting works on watchOS the same way it does on iOS — which is important for the channel multiplexing argument.

Your observation that the G7's Direct-To-Watch support for the ONE+ is already in the IPA behind geofencing is a fascinating side note — I wonder if that same entitlement pattern applies there too.

Thanks for the extraordinary work documenting all of this.

**Reply by gui-dos** [reply]
@cachrisman:

No, I have never tried those BLE background modes and requested those restricted entitlements by contacting Apple directly: perhaps thanks to that my TestFlight beta has not been taken down yet… ;-)

It still includes my standalone Watch app, and yep, I could really not only intercept the incoming decrypted BLE data directly on the wrist but also request to backfill it by simply using the opcode `0x59` (it requires just to append eight bytes, the start and end timestamps) before the Vendor’s app disconnected, as you can see on <https://github.com/gui-dos/DiaBLE/blob/cfc3a94827a07ff3035b39e35ea04f00d8f34c16/DiaBLE/DexcomG7.swift#L293-L298> (the independent Apple Watch app and iPad playground make use of the same code as the iPhone app).

**Comment by MechaSteve** [comment]
Very cool work here.

I have a working implementation of a standalone G6 receiver implemented on an ESP32-S3 based system.

I am working on extending the system to support the G7 sensors, and would welcome any insight you can give. 

If I understand your explanation, there is significantly more security encryption and handshakes involved with the G7 to complete pairing, compared to the G6. Is that correct, or is the additional security limited to the more advanced functions?

**Reply by MechaSteve** [reply]
https://github.com/MechaSteve/Flying-CGM

**Comment by gui-dos** [comment]
I noticed in https://github.com/j-kaltes/Juggluco/blob/f9aad1a3080e92c42e253b96ed511cb7ba5ac5b2/Common/src/dex/java/tk/glucodata/DexGattCallback.java#L774-L779 that Juggluco is using the same certificate data that xDrip+ is requiring to QR-scan manually (https://navid200.github.io/xDrip/docs/Dexcom/G7.html) but unfortunately these days I am not experimenting actively with Dexcoms since their Direct-To-Watch implementation is already optimal to try to tweak and improve it…

Yep: the J-PAKE authentication protocol is now required to generate the ephemeral keys. I am seeing that all the other iOS partners' apps are wrapping the `mbedtls` framework but I am desisting to try to port it since my project got little feedback from the community about the much richer set of `opcodes` commands availlaBLE with the G7… :-(
